platform/engineering/design-docs/2021-08-01-autonomous-deployments.md

```diff
@@ -0,0 +1,220 @@
+
+A representative from each VFS team should have to review / approve the PR that activated this feature in each environment as a way to acknoledge the change. FE Tools will also attend the VSA Engineering meeting, alert slack, and add to the newsletter (with each environment release).
+
+This release behavior should not replace Review Instances. Review Instances should still be used when possible as to not overload staging / dev deployments.
+
+#### Milestone: Create and release initial production behaviors with "allow" list
+
+- Create "allow" list for applications
+- Create production deployment workflow
+- Monitor auto and manual deployments for a week to debug and troubleshoot
+
+This work should be completed in 2-3 sprints with 2 people, one of which should be very familiar with Github Actions.
+
+#### Milestone: Create and release MVP production behaviors with "disallow" list
+
+- Update deployment workflow to use disallow list instead of allow list
+- Plan and execute VFS training on the update
+- Release to teams
+- Monitor auto and manual deployments for a sprint to debug and troubleshoot
+
+This work should be completed in 2-3 sprints with 2 people.
+
+### Alternatives
+
+An alternative that was discussed was isolating the applications in their own repositories before executing autonomous deploys. We felt that this would bloat the timeline and delay the feature from being released for too long. Isolating each application entails all of the groundwork needed to complete the autonomous deployment initiative, but also revamping deployments further, dependency isolation, foundational upgrades to the infrastructure, and large/disruptive changes to VFS teams' workflow as well as extensive training. While full isolation is a long term goal, it will be rolled out as part of an iterative plan; autonomous deployments being the first objective to complete.
+
+We also discussed monorepo isolation, but deemed it as unnecessary to complete the objective. Monorepo setup involves a technology infusion with something like Lerna, Yarn Workspaces, and/or NX. It was decided that pulling in new infrastructure at this stage would not get us any closer to the primary objective. This was also identified as an important upgrade to the system, but as it is non-critical, it is to be added to the roadmap as a follow-up initiative. Before this work can be completed, we need to make foundational upgrades to the platform code, architecture, and design system.
+
+One last option was to leave the system as-is, which the current deployment model. This was not pursued because it would not meet the objective. Another option was to keep everything we have in place right now, but require full builds and deploys for each merge into master. Ultimately we felt the deployment process would be too slow and prone to failure if we required full build and test suites for each deployment, which would dramatically increase strain on support.
```

Comment:
I'd like to see another alternative considered:
- Use only `master` to deploy to all three environments
- In the `master` CI pipeline, only build, test, and deploy applications affected by changes since the last deploy
- Deploy to all environments at once
- Employ [feature toggles](https://www.martinfowler.com/articles/feature-toggles.html) to separate deploys from feature releases